Allow rolling-update of a single container in multi-container pods

munnerz opened the pull request:
This is a fix for #13483

It allows for the updating of a single containers image within a pod with multiple containers, in order to make rolling updates on multi-container pods easier.

Commit:
Allow rolling-update of a single container in multi-container pods

Accept codec as parameter to CreateNewControllerFromCurrentController function. Add tests for performing a rolling update on a container in a multi-container pod.

munnerz commented:
I signed it!

Signed as ASAP Compare Ltd.

mikedanese commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -539,13 +540,29 @@ func CreateNewControllerFromCurrentController(c *client.Client, namespace, oldNa
+		}
+
+		if !containerFound {
+			return nil, goerrors.New(fmt.Sprintf("Container %s not found in pod", container))
```
Use fmt.Errorf() instead of goerrors.New(fmt.Sprintf(

mikedanese commented:
What happens when someone uses --container=name with a name not in the pod but there is only one container in the pod?

mikedanese commented:
cc @bgrant0607 @janetkuo @kubernetes/kubectl

munnerz commented:
Thanks for the comments.

I've updated to account for specifying the container name when there's only one container in the pod (so if the user now specifies a container, it must exist within the pod regardless of the number of containers in the pod).

Rebased & regenerated docs based off 1.1.

kargakis commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -531,21 +531,38 @@ func LoadExistingNextReplicationController(c client.ReplicationControllersNamesp
-	if len(newRc.Spec.Template.Spec.Containers) > 1 {
-		// TODO: support multi-container image update.
-		return nil, goerrors.New("Image update is not supported for multi-container pods")
+	if container != "" {
```
We follow a convention of comparing the length of a string against zero instead of an empty string. So this should be `len(container) != 0`.

kargakis commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -531,21 +531,38 @@ func LoadExistingNextReplicationController(c client.ReplicationControllersNamesp
+		}
+
+		if !containerFound {
+			return nil, fmt.Errorf("Container %s not found in pod", container)
```
lowercase the error

munnerz replied:
The other errors (which using goerrors.New) are mixed case - should I also change those to lowercase at the same time or does goerrors handle that?

kargakis replied:
hm, not sure about goerrors, but whatever comes out of errors.New or fmt.Errorf should be lowercased

kargakis commented on pkg/kubectl/rolling_updater.go:
```diff
@@ -531,21 +531,38 @@ func LoadExistingNextReplicationController(c client.ReplicationControllersNamesp
+		}
+	}
+
+	if len(newRc.Spec.Template.Spec.Containers) > 1 && container == "" {
```
`len(container) == 0`

kargakis commented:
a couple of nits, lgtm otherwise

also a unit test would be appreciated

bgrant0607 commented:
Yes, please add a test.

bgrant0607 commented:
@k8s-bot unit test this

lachie83 commented:
What's the status of getting this merged.?There is no way to do rolling updates on multi-container pods which breaks the side-car container model which is a strength of pods in Kubernetes

mikedanese commented:
@lachie83 Yes there is. You just have to do it with an object config file rather than with command line options. As stated above we are waiting for the PR author to add a unit test.

munnerz commented:
I'll try and get a unit test done this evening - I've been looking over
existing unit tests to work out how best to structure my own.

Yeah it is currently possible, however tedious as it involves providing a
new RC definition, replacing particular parts (version tags, RC name) and
renaming the new RC to match the old one after success (talking from an
automated CI workflow standpoint).

On Thursday, 3 December 2015, Mike Danese <notifications@github.com> wrote:

> @lachie83 <https://github.com/lachie83> Yes there is. You just have to do
> it with an object config file rather than with command line options. As
> stated above we are waiting for the PR author to add a unit test.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17111#issuecomment-161741843>
> .
>

lachie83 commented:
Correct me if I'm wrong but the method you are referencing (kubectl update/replace rc?) will not do a "rolling-update". It will simply reconfigure the existing rc with the new images such that you loose the ability to roll out one pod at a time and re-attach it to the service after heathchecks have passed.

munnerz commented:
No, rolling-update can also accept a file that defines the new RC to update
to. You don't need to run kubectl replace.

kubectl rolling-update rc_name -f new_rc.yml

On Thursday, 3 December 2015, Lachlan Evenson <notifications@github.com>
wrote:

> Correct me if I'm wrong but the method you are referencing (kubectl
> update/replace rc?) will not do a "rolling-update". It will simply
> reconfigure the existing rc with the new images such that you loose the
> ability to roll out one pod at a time and re-attach it to the service after
> heathchecks have passed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17111#issuecomment-161745644>
> .
>

mikedanese commented:
@lachie83 
```
# write an rc
$ vi rc.yml
# create rc
$ kubectl create -f rc.yml
# modify rc to update one image tag (you need to update the name, pod template labels, and the selector so it doesn't match the old rc)
$ vi rc.yml
# preform rolling update 
$ kubectl rolling-update <old-rc-name> -f rc.yml
```
Agree that the UX introduced by this PR is much improved but a rolling update _is_ possible today on multi container pods. Also checkout the Deployment object.

munnerz commented:
The major UX issue I experienced with the above workflow is the fact that
the RC to update to cannot have the same name as the old RC, meaning it's
the responsibility of the user to manually rename the old RC after the
update else the next run will not succeed.

Deployments do provide a cleaner interface for updating RCs, however add
extra complexity for users wanting to quickly update a RC.

On Thursday, 3 December 2015, Mike Danese <notifications@github.com> wrote:

> @lachie83 <https://github.com/lachie83>
>
> # write an rc
> $ vi rc.yml
> # create rc
> $ kubectl create -f rc.yml
> # modify rc to update one image tag (you need to update the name, pod template labels, and the selector so it doesn't match the old rc)
> $ vi rc.yml
> # preform rolling update
> $ kubectl rolling-update <old-rc-name> -f rc.yml
>
> Agree that the UX introduced by this PR is much improved but a rolling
> update *is* possible today on multi container pods. Also checkout the
> Deployment object.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17111#issuecomment-161747005>
> .
>

lachie83 commented:
@mikedanese @munnerz - Thanks guys. I understand what you're saying. How are you doing that in an automated fashion off a CI deploy loop? Inserting the build number in the name and all the other labels that need to be changed in the rc somehow?

This is what we're currently doing on single container pods on a CI deploy

`kubectl --namespace=hello rolling-update --update-period="1s" hello --image=lachie83/hello:latest`

mikedanese commented:
@munnerz not to mention deployments are still in alpha :smile:

munnerz commented:
@lachie83 

We currently use the following (rather convoluted) setup:

```bash
sed -e 's?{CI_HASH}?'"${TRAVIS_COMMIT:0:7}"'?g' --in-place build/replication-controller.json
cp build/replication-controller.json build/replication-controller-new.json
sed -e 's?{RC_NAME}?'"$RC_NAME"'?g' --in-place build/replication-controller.json
sed -e 's?{RC_NAME}?'"$RC_NAME_NEW"'?g' --in-place build/replication-controller-new.json

kubectl rolling-update $RC_NAME -f build/replication-controller-new.json
kubectl delete rc --cascade=false "$RC_NAME_NEW"
kubectl create -f build/replication-controller.json
```

mikedanese commented:
@lachie83 deployment uses a generated name. https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L76, if you create an object without a .metadata.name and specify .metadata.generatedName = "some-controller-" the apiserver will populate it's name with a unique string that looks like "some-controller-XXXX". Putting version, or hash of pod template also make since in place of the random suffix.

mtougeron commented:
One of the concerns we have with this method though is that it will replace ALL containers running in the RC not just the one(s) that should be updated. e.g., in our case we have 2 containers in an RC, one is our application and the other is a DataDog sidecar. We want to update the container version that the DataDog sidecar is running but we don't want to also reload the application container. Won't the method you are proposing always do both containers in the RC?

mikedanese commented:
@mtougeron That's a problem with or without the change in this PR. Both will replace the entire pod.

mtougeron commented:
Doh :(

mtougeron commented:
@mikedanese @munnerz Thanks for all the answers to our questions!

lachie83 commented:
+1 @mikedanese @munnerz

munnerz commented:
In order to test the CreateNewControllerFromCurrentController function, I need a testclient that has a Codec field as the function depends upon it for generating the deployment hash.

Is there an alternative test client to testclient.Fake that can be used?

kargakis commented:
You need to use a bare fake client and prepend a reaction to the only call inside the unit you are testing (GET replicationcontroller/oldname) - [example](https://github.com/openshift/origin/blob/693be191a98228377df5e6fe4a747bc993018452/pkg/cmd/cli/cmd/deploy_test.go#L132-L135). By controlling the output of the action you can test various scenarios).

kargakis commented:
For the codec, use [testapi.Codec](https://github.com/kubernetes/kubernetes/blob/a5100ef05726237e4e1f44997852963c955b608c/pkg/api/testapi/testapi.go#L103).

munnerz commented:
@kargakis I've added tests for the CreateNewControllerFromCurrentController function. Hopefully they're acceptable :smile:

kargakis commented:
I'm out today, @bgrant0607 can you re-assign to someone else for reviewing? Or I'll have a look on Monday.

jlowdermilk commented:
lgtm, thanks for adding tests. Please squash commits and I'll add label.

munnerz commented:
Squashed :smile:

Thanks for reviewing!

munnerz commented:
Mesos tests seem to be failing due to:

`[18:49:23][Step 6/6] ea3ced1122e1: Error pulling image (v2.2.1) from quay.io/coreos/etcd, ApplyLayer exit status 1 stdout:  stderr: write /etcd: no space left on device`

jlowdermilk commented:
Re corp cla, I don't see it on list, looking into it.

jlowdermilk commented:
@munnerz, checked with someone internally. They asked to try again via https://cla.developers.google.com/clas

munnerz commented:
@jlowdermilk Got it signed again yesterday :smile:

jlowdermilk commented:
@munnerz thanks, cla is on file. You might need to reply "I signed it" again to trigger the bot. Also, the commit email needs to match the one associated with the corp cla/contributor group.

munnerz commented:
I signed it

jlowdermilk commented:
Thanks @munnerz, sorry about the long wait.

munnerz commented:
No worries, thanks for your help - any idea what release this'll be rolled into?

jlowdermilk commented:
We typically only do cherrypick/patch releases for bug fixes, so this will get picked up by the next minor release, which would be 1.2.

mikedanese commented:
@k8s-bot ok to test

jlowdermilk commented:
@k8s-bot test this

janetkuo commented:
@k8s-bot e2e test this

mikedanese commented:
@k8s-bot ok to test
